### ISSUE TITLE
DTLS handshake failure happens when a JSSIP Web agent calls to a legacy SIP client through Kamailio SIP Proxy

### ISSUE THREAD

**Issue by BroWater** [sipwise/rtpengine#741]
Hi all,

I use the JSSIP library to build  a JSSIP Web agent and make a call described as title,
however, I couldn't see the video on both sites,
and I got the logs below:

Mar 21 17:09:14 rtpengine[21323]: INFO: [u4hf26pi7hcebi33etf5 port 30390]: ICE negotiated: local interface 10.100.92.52
Mar 21 17:09:14 rtpengine[21323]: INFO: [u4hf26pi7hcebi33etf5 port 30422]: ICE negotiated: local interface 10.100.92.52
Mar 21 17:09:14 rtpengine[21323]: ERR: [u4hf26pi7hcebi33etf5 port 30390]: DTLS error: 1 (sslv3 alert handshake failure)
Mar 21 17:09:14 rtpengine[21323]: ERR: [u4hf26pi7hcebi33etf5 port 30390]: DTLS error on local port 30390
Mar 21 17:09:14 rtpengine[21323]: ERR: [u4hf26pi7hcebi33etf5 port 30422]: DTLS error: 1 (sslv3 alert handshake failure)
Mar 21 17:09:14 rtpengine[21323]: ERR: [u4hf26pi7hcebi33etf5 port 30422]: DTLS error on local port 30422

I have no idea now since I do not understand DTLS well.
Any suggestions?

Version: git-master-63bdd61
OpenSSL version: 1.0.1g
Kamailio version: 5.2.1

Or maybe this is a issue of Kamailio?!

If any further information needed, please let me know.
Thanks. :)

**Comment by rfuchs** [comment]
Hard to say - OpenSSL error reporting isn't all that great. Maybe try a newer OpenSSL version?
Other than that, it might be something that your JSSIP client is doing wrong. You can try look at a pcap in Wireshark to see if that reveals anything.
It's definitely not related to Kamailio.

**Comment by BroWater** [comment]
I upgrade OpenSSL to 1.0.2r (latest ) and recompile TLS module from Kamailio and RTPEngine.

Now when I make the call from JSSIP client, RTPEngine suddenly crashed.

I dump the log to foreground it showed that:

[1553291232.752496] INFO: Generating new DTLS certificate
[1553291233.073832] INFO: Startup complete, version git-master-63bdd61


[1553291314.734283] INFO: Received command 'ping' from 127.0.0.1:48232
[1553291314.734381] INFO: Replying to 'ping' from 127.0.0.1:48232
[1553291417.681890] INFO: Received command 'ping' from 127.0.0.1:47239
[1553291417.681952] INFO: Replying to 'ping' from 127.0.0.1:47239
[1553291454.167578] INFO: [d87d2jp85bv58tr1s7r3]: Received command 'offer' from 127.0.0.1:37647
[1553291454.168691] NOTICE: [d87d2jp85bv58tr1s7r3]: Creating new call
./rtpengine: symbol lookup error: ./rtpengine: undefined symbol: DTLS_client_method

Maybe this means I didn't really use the version 1.0.2r  of OpenSSL??

**Comment by rfuchs** [comment]
That looks like a broken OpenSSL installation. Perhaps you have multiple conflicting versions of the OpenSSL libraries installed? Also did you recompile everything with the new development headers present?

**Comment by BroWater** [comment]
Sorry for the late reply ~

The problem has been solved and here comes the conclusion.

Back to the earlier question - DTLS handshake error，
I tested the JSSIP client on Chrome and FireFox  while the latter  works,
it seems that Chrome has a strict requirement of DTLS_v1.2 but FireFox does not so far,
Then I renewed the OpenSSL to version 1.0.2r which supports DTLS_v1.2,
(A bit of struggle here since I have multiple toolchains and OpenSSLs installed on my development environment , I even have to adjust pkg-config command in daemon/Makefile for indicating the right prefix of packages needed)
and the updated RTPEngine crashed during connection between a legacy SIP and JSSIP client:

> "./rtpengine: symbol lookup error: ./rtpengine: undefined symbol: DTLS_client_method" 

The reason is that the older ssl and crypto dynamic libraries (.so)  were put under the same directory with kamailio and RTPEngine and not been updated.

Anyway, thanks for the help. 
:)
